### ISSUE TITLE
debugging gradients in a package doesn't work when in debugging mode

### ISSUE THREAD

**Issue by njtierney** [rstudio/tensorflow#539]
Hello!

We are trying to debug an issue with `greta` where we were getting a "none" gradient:

```
Error in py_call_impl(callable, dots$args, dots$keywords) : 
  RuntimeError: Evaluation error: ValueError: Encountered `None` gradient.
  fn_arg_list: [<tf.Tensor 'free_state:0' shape=(4, 1) dtype=float64>]
  grads: [None]
```

We tried to explore the error using gradient tape, whilst in a debugging mode.

However GradientTape wasn't working - we have a small reprex below, demonstrating that while in debugging mode, you cannot use GradientTape. We think it might have something to do with the package being loaded in the same way as in `library(tensorflow)`, but we're not quite sure how to solve that in pkg development context, since the code we're trying to fix can't be replicated outside of the package.

If you enter debug mode as shown below, then we get the error, 

```
Error in py_resolve_dots(list(...)) : object 'z' not found
```

```r
library(greta)
debug(greta::normal)
x <- normal(0, 1)

# in debug mode
library(tensorflow)

x <- tf$ones(shape(2, 2))

with(tf$GradientTape() %as% t, {
  t$watch(x)
  y <- tf$reduce_sum(x)
  z <- tf$multiply(y, y)
})

# Derivative of z with respect to the original input tensor x
dz_dx <- t$gradient(z, x)
```

Below is my session info, if that helps

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value
#>  version  R version 4.2.0 (2022-04-22)
#>  os       macOS Monterey 12.3.1
#>  system   aarch64, darwin20
#>  ui       X11
#>  language (EN)
#>  collate  en_AU.UTF-8
#>  ctype    en_AU.UTF-8
#>  tz       Australia/Perth
#>  date     2022-06-16
#>  pandoc   2.17.1.1 @ /Applications/RStudio.app/Contents/MacOS/quarto/bin/ (via rmarkdown)
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version    date (UTC) lib source
#>  cli           3.3.0.9000 2022-06-15 [1] Github (r-lib/cli@31a5db5)
#>  crayon        1.5.1      2022-03-26 [1] CRAN (R 4.2.0)
#>  digest        0.6.29     2021-12-01 [1] CRAN (R 4.2.0)
#>  ellipsis      0.3.2      2021-04-29 [1] CRAN (R 4.2.0)
#>  evaluate      0.15       2022-02-18 [1] CRAN (R 4.2.0)
#>  fansi         1.0.3      2022-03-24 [1] CRAN (R 4.2.0)
#>  fastmap       1.1.0      2021-01-25 [1] CRAN (R 4.2.0)
#>  fs            1.5.2      2021-12-08 [1] CRAN (R 4.2.0)
#>  glue          1.6.2      2022-02-24 [1] CRAN (R 4.2.0)
#>  highr         0.9        2021-04-16 [1] CRAN (R 4.2.0)
#>  htmltools     0.5.2      2021-08-25 [1] CRAN (R 4.2.0)
#>  knitr         1.39       2022-04-26 [1] CRAN (R 4.2.0)
#>  lifecycle     1.0.1      2021-09-24 [1] CRAN (R 4.2.0)
#>  magrittr      2.0.3      2022-03-30 [1] CRAN (R 4.2.0)
#>  pillar        1.7.0      2022-02-01 [1] CRAN (R 4.2.0)
#>  pkgconfig     2.0.3      2019-09-22 [1] CRAN (R 4.2.0)
#>  purrr         0.3.4      2020-04-17 [1] CRAN (R 4.2.0)
#>  R.cache       0.15.0     2021-04-30 [1] CRAN (R 4.2.0)
#>  R.methodsS3   1.8.1      2020-08-26 [1] CRAN (R 4.2.0)
#>  R.oo          1.24.0     2020-08-26 [1] CRAN (R 4.2.0)
#>  R.utils       2.11.0     2021-09-26 [1] CRAN (R 4.2.0)
#>  reprex        2.0.1      2021-08-05 [1] CRAN (R 4.2.0)
#>  rlang         1.0.2      2022-03-04 [1] CRAN (R 4.2.0)
#>  rmarkdown     2.14       2022-04-25 [1] CRAN (R 4.2.0)
#>  rstudioapi    0.13       2020-11-12 [1] CRAN (R 4.2.0)
#>  sessioninfo   1.2.2      2021-12-06 [1] CRAN (R 4.2.0)
#>  stringi       1.7.6      2021-11-29 [1] CRAN (R 4.2.0)
#>  stringr       1.4.0      2019-02-10 [1] CRAN (R 4.2.0)
#>  styler        1.7.0      2022-03-13 [1] CRAN (R 4.2.0)
#>  tibble        3.1.7      2022-05-03 [1] CRAN (R 4.2.0)
#>  utf8          1.2.2      2021-07-24 [1] CRAN (R 4.2.0)
#>  vctrs         0.4.1      2022-04-13 [1] CRAN (R 4.2.0)
#>  withr         2.5.0      2022-03-03 [1] CRAN (R 4.2.0)
#>  xfun          0.31       2022-05-10 [1] CRAN (R 4.2.0)
#>  yaml          2.3.5      2022-02-21 [1] CRAN (R 4.2.0)
#> 
#>  [1] /Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/library
#> 
#> ──────────────────────────────────────────────────────────────────────────────
```

</details>

Also for further context, we are working on this branch of greta - https://github.com/njtierney/greta/tree/tf2-poke-tf-fun

specifically, putting a browser here: https://github.com/njtierney/greta/blob/tf2-poke-tf-fun/R/inference_class.R#L705

And running this code:

```r
devtools::load_all()
x <- normal(0, 1)
m <- model(x)
mcmc(m)
```

We then get the above error,

```
Error in py_call_impl(callable, dots$args, dots$keywords) : 
  RuntimeError: Evaluation error: ValueError: Encountered `None` gradient.
  fn_arg_list: [<tf.Tensor 'free_state:0' shape=(4, 1) dtype=float64>]
  grads: [None]
```

If that extra info helps!

**Comment by t-kalinowski** [comment]
Hi, 

I installed the dev version of greta from github, ran `install_greta_deps()`, then ran the smaller reprex and it succeeded without error for me. This was the output I got:

```
Restarting R session...

> library(greta)

Attaching package: ‘greta’

The following objects are masked from ‘package:stats’:

    binomial, cov2cor, poisson

The following objects are masked from ‘package:base’:

    %*%, apply, backsolve, beta, chol2inv, colMeans, colSums, diag, eigen, forwardsolve,
    gamma, identity, rowMeans, rowSums, sweep, tapply

> debug(greta::normal)
> x <- normal(0, 1)
debugging in: normal(0, 1)
debug: {
    distrib("normal", mean, sd, dim, truncation)
}
Browse[2]> n
debug: distrib("normal", mean, sd, dim, truncation)
Browse[2]> n
✔ Initialising python and checking dependencies ... done!               

exiting from: normal(0, 1)
> x
greta array (variable following a normal distribution)

     [,1]
[1,]  ?  
> 
> # in debug mode
> library(tensorflow)
> 
> x <- tf$ones(shape(2, 2))
> 
> with(tf$GradientTape() %as% t, {
+     t$watch(x)
+     y <- tf$reduce_sum(x)
+     z <- tf$multiply(y, y)
+ })
> 
> # Derivative of z with respect to the original input tensor x
> dz_dx <- t$gradient(z, x)
> dz_dx
Tensor("Tile:0", shape=(2, 2), dtype=float32)
> tf_config()
TensorFlow v1.14.0 (~/.local/share/r-miniconda/envs/greta-env/lib/python3.7/site-packages/tensorflow)
Python v3.7 (~/.local/share/r-miniconda/envs/greta-env/bin/python)
> sessionInfo()
R version 4.2.0 (2022-04-22)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.4 LTS

Matrix products: default
BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/libmkl_rt.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] tensorflow_2.9.0.9000 greta_0.4.2.9000     

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.8.3         rstudioapi_0.13      parallelly_1.32.0    magrittr_2.0.3      
 [5] whisker_0.4          rappdirs_0.3.3       hms_1.1.1            progress_1.2.2      
 [9] here_1.0.1           lattice_0.20-45      R6_2.5.1             rlang_1.0.2         
[13] globals_0.15.0       tools_4.2.0          parallel_4.2.0       grid_4.2.0          
[17] png_0.1-7            cli_3.3.0            coda_0.19-4          tfruns_1.5.0        
[21] ellipsis_0.3.2       rprojroot_2.0.3      digest_0.6.29        lifecycle_1.0.1     
[25] crayon_1.5.1         processx_3.6.0       Matrix_1.4-1         callr_3.7.0         
[29] base64enc_0.1-3      vctrs_0.4.1          ps_1.7.0             codetools_0.2-18    
[33] glue_1.6.2           compiler_4.2.0       prettyunits_1.1.1    reticulate_1.25-9000
[37] jsonlite_1.8.0       future_1.26.1        listenv_0.8.0        pkgconfig_2.0.3   
```

**Comment by t-kalinowski** [comment]
I ran the tf2 branch you pointed me to and was able to reproduce the error 🎉 . It looks like the error in your code is originating from `tfp$mcmc$sample_chain`, and you're not actually calling `tf$GradientTape()` directly. Is that correct?

```r
Error in py_call_impl(callable, dots$args, dots$keywords) : 
  ValueError: Encountered `None` gradient.
  fn_arg_list: [<tf.Tensor: shape=(4, 1), dtype=float64, numpy=
array([[-0.06387527],
       [-0.03749587],
       [ 0.0033471 ],
       [ 0.07576335]])>]
  grads: [None]
Browse[1]> reticulate::py_last_error()
Traceback (most recent call last):
  File "/home/tomasz/.virtualenvs/r-reticulate/lib/python3.9/site-packages/tensorflow_probability/python/mcmc/sample.py", line 330, in sample_chain
    previous_kernel_results = kernel.bootstrap_results(current_state)
  File "/home/tomasz/.virtualenvs/r-reticulate/lib/python3.9/site-packages/tensorflow_probability/python/mcmc/hmc.py", line 561, in bootstrap_results
    kernel_results = self._impl.bootstrap_results(init_state)
  File "/home/tomasz/.virtualenvs/r-reticulate/lib/python3.9/site-packages/tensorflow_probability/python/mcmc/metropolis_hastings.py", line 273, in bootstrap_results
    pkr = self.inner_kernel.bootstrap_results(init_state)
  File "/home/tomasz/.virtualenvs/r-reticulate/lib/python3.9/site-packages/tensorflow_probability/python/mcmc/hmc.py", line 775, in bootstrap_results
    ] = mcmc_util.maybe_call_fn_and_grads(self.target_log_prob_fn, init_state)
  File "/home/tomasz/.virtualenvs/r-reticulate/lib/python3.9/site-packages/tensorflow_probability/python/mcmc/internal/util.py", line 306, in maybe_call_fn_and_grads
    raise ValueError('Encountered `None` gradient.\n'
ValueError: Encountered `None` gradient.
  fn_arg_list: [<tf.Tensor: shape=(4, 1), dtype=float64, numpy=
array([[-0.06387527],
       [-0.03749587],
       [ 0.0033471 ],
       [ 0.07576335]])>]
  grads: [None]

```

Looking around that stack track it looks like the actual error is emitted from https://www.tensorflow.org/probability/api_docs/python/tfp/math/value_and_gradient, called by https://www.tensorflow.org/probability/api_docs/python/tfp/mcmc/HamiltonianMonteCarlo#bootstrap_results.

I can't tell without digging further if this is a TFP bug or a greta bug. `None` values for gradients are usually due to disconnected tensors. E.g.:
```r
x <- tf$Variable(2)
y <- tf$Variable(3)

with(tf$GradientTape() %as% tape, {
  z <- y * y
})
tape$gradient(z, x)
```
This section in the Python docs lists 4 common causes of a `None` gradient: https://www.tensorflow.org/guide/autodiff#getting_a_gradient_of_none
It might be helpful for troubleshooting.

Hope this helps a little!

**Comment by njtierney** [comment]
Hi @t-kalinowski !

Thanks so much for going through installing greta and everything to replicate the bug!

>  It looks like the error in your code is originating from tfp$mcmc$sample_chain, and you're not actually calling tf$GradientTape() directly. Is that correct?

Yes - we don't call `tf$GradientTape()` inside greta, but we were hoping to use it to help debug the error, with something like the following at that line: https://github.com/njtierney/greta/blob/tf2-poke-tf-fun/R/inference_class.R#L705

```r
x <- as_tensor(free_state)

with(
  tf$GradientTape() %as% tape, {
    tape$watch(x)
    y <- self$model$dag$tf_log_prob_function_adjusted(x)
  }
)
dy_dx <- tape$gradient(y, x)

```

However when we try to use the debugger there it doesn't work when we're inside `browser()` - would you expect that to work?

Thank you very much again for your help, and for the links! We'll take a read of the common causes of a None gradient now, hopefully that sheds some light! Will let you know how we get on.

**Comment by t-kalinowski** [comment]
Hi, I wouldn't expect a `browser()` to interfere with anything. You should be able to safely drop into an R debugger while a Python context is open, including `tf$GradientTape()` contexts, or even tracing of a `tf_function()` or `dataset_map()`. 

E.g., this should work:
```r
with(tf$GradientTape() %as% tape, { browser(), ... })
```

**Comment by t-kalinowski** [comment]
One thing to be aware of though is that you have to exit the debugger by continuing, not quitting. If you exit by quitting, then the Python context is never closed and you can only recover by restarting the R session. In a long-lived interactive session, you can repeatedly enter and exit from a debugging `browser()` inside a Python context, so long as you "exit" the browser by letting it resume regular execution, which means letting the R interpreter invoke the `` gradient_tape$`__exit__`() `` method.

**Comment by njtierney** [comment]
Hi @t-kalinowski !

Thanks again for your help, really appreciate it. What we are wanting to do is evaluate the gradient tape inside browser, not browser inside the gradient tape environment.

We suspect that there might be some issue with how dimensions or something like that are passed around in greta, and so what we want to do is check if we can run `tf$GradientTape` on the inputs we are about to pass to TFP. 

We want to put a browser in the package code so that we can interactively check gradients as the mode is being defined
(we put it at https://github.com/njtierney/greta/blob/tf2-poke-tf-fun/R/inference_class.R#L705 if that helps). But we need to be able to run gradient tape in there to do that.

When we run even a simple version of gradient tape from the examples, we get an error we can't get past. e.g. if you initiate a debugger (and don't exit it), then run the following code while in debug mode, you can hopefully reproduce the error:


```r
# hit `n` to take one step from browser
# now try and run GradientTape on the inputs
# load tensorflow to get access to `%as%` method
library(tensorflow)
# Attaching package: ‘tensorflow’
#
# The following object is masked from ‘package:greta’:
#
# tf
x <- as_tensor(free_state)
x
# tf.Tensor(
# [[-0.06387527]
# [-0.03749587]
# [ 0.0033471 ]
# [ 0.07576335]], shape=(4, 1), dtype=float64)
with(
tf$GradientTape() %as% tape, {
tape$watch(x)
y <- self$model$dag$tf_log_prob_function_adjusted(x)
}
)
# debug at #3: tape$watch(x)
dy_dx <- tape$gradient(y, x)
# Error in py_resolve_dots(list(...)) : object 'y' not found
# Called from: py_resolve_dots(list(...))
```


Note that when you ran the reprex before, you exited the browser before running the code, so it worked. We need to run it whilst still in the browser.

**Comment by t-kalinowski** [comment]
Hi, all a `browser()` call does is interrupt execution of R code, and give you an interactive console with some additional commands for controlling how the next steps of the interpreter should run. So when you say:

> What we are wanting to do is evaluate the gradient tape inside browser, not browser inside the gradient tape environment

A `browser()` is just a breakpoint. Like a "point" in geometry, they can exist in a space, but they occupy no space. It's not an environment, in this sense of the word.

All `reticulate:::with.python.buildin.object()` does is call ``tape$`__enter__`() ``, evaluate the 2nd argument to `with()`, and then call `` `tape$`__exit__`() ``. `tape %as% t` also makes the symbol `t` available in the same environment. You can insert a breakpoint (with `browser()`) at any and all stages in this sequence: before the scope is opened, anytime it remains open, or after it's been closed.
```r
browser()
with(tf$GradientTape() %as% tape, { browser(); ...; browser() })
browser()
```

If, when you drop into a debugger, it appears there are missing symbols that you expect to be there, then that's more likely that you've interrupted execution in the wrong R environment, or code is evaluating in the wrong R environment. `with()` evaluates its `expo` in the same environment as where it's called, so it's unlikely to be the cause of missing symbols. (e.g., in the snippet above, each of the `browser()` called will drop you into the __same__ R environment). Gradients of course won't be available until after the `GradientTape()` scope is closed.

https://user-images.githubusercontent.com/8462255/174839078-f1ebe788-2b1e-42c9-ada2-03027bec5adc.mov

Does this help?

**Comment by njtierney** [comment]
Hi @t-kalinowski - thank you so much for your help here!

Looking a bit closer at our debugging process, we realised that for whatever reason, the debugger was stepping one layer deeper than the environment where the variables were defined when we were debugging.

E.g., we would be in `browser[1]` and then when we were in the `with` call, it stepped into `browser[2]`, and didn't know about the existence of the variable we wanted to differentiate.

So, consider this user error - apologies! We worked out how to get gradients now and then debug greta, we've now got a set of different problems to consider - which is good news!

Thanks again for your patience and help with this, using a video reprex between Nick Golding and I helped us understand this problem, so I'll be borrowing that idea into the future!

Cheers,

Nick
